lib/ide/src/controller.rs

```diff
@@ -0,0 +1,426 @@
+/// methods of their own. While not necessary, such implementation may allow
+/// hiding from user gritty details of `with` usage behind nice, easy API.
+macro_rules! make_handles {
+    ($data_type:ty) => {
+        /// newtype wrapper over StrongHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct StrongHandle(pub utils::cell::StrongHandle<$data_type>);
+
+        impl StrongHandle {
+            /// Obtain a WeakHandle to this data.
+            pub fn downgrade(&self) -> WeakHandle {
+                WeakHandle(self.0.downgrade())
+            }
+            /// Create a new StrongHandle that will wrap given data.
+            pub fn new(data:$data_type) -> StrongHandle {
+                StrongHandle(utils::cell::StrongHandle::new(data))
+            }
+        }
+
+        /// newtype wrapper over WeakHandle.
+        #[derive(Shrinkwrap)]
+        #[derive(Clone,Debug)]
+        pub struct WeakHandle(pub utils::cell::WeakHandle<$data_type>);
+
+        impl WeakHandle {
+            /// Obtain a StrongHandle to this data.
+            pub fn upgrade(&self) -> Option<StrongHandle> {
+                self.0.upgrade().map(StrongHandle)
+            }
+        }
+    };
+}
+
+
+
+// =======================
+// === Text controller ===
+// =======================
+
+pub mod project {
+    //! Project controller.
+    ///
+    /// Responsible for owning any remote connection clients. Expected to live
+    /// as long as the project remains open in the IDE.
+    use super::*;
+    use json_rpc::Transport;
+
+    /// Create a new project controller.
+    ///
+    /// The remote connections should be already established.
+    pub fn new(file_manager_transport:impl Transport + 'static) -> StrongHandle {
+        let file_manager = file_manager_client::Client::new(file_manager_transport);
+        let ret = Data {
+            file_manager,
+            module_cache: default(),
+        };
+        StrongHandle::new(ret)
+    }
+
+    /// Project controller's state.
+    #[derive(Debug)]
+    pub struct Data {
+        /// File Manager Client.
+        file_manager : file_manager_client::Client,
+        /// Cache of module controllers. As the weak is handle (we don't manage
+        /// their lifetime), dead ones might still be included.
+        module_cache : HashMap<module::Location, module::WeakHandle>,
+    }
+
+    impl Data {
+        /// Returns handle to given module's controller if already present.
+        pub fn lookup_module(&mut self, loc:&module::Location) -> Option<module::StrongHandle> {
+            let weak = self.module_cache.get(loc)?;
+            match weak.upgrade() {
+                Some(strong) => Some(strong),
+                None => {
+                    self.module_cache.remove(loc);
+                    None
+                }
+            }
+        }
+
+        /// Stores given module controller handle and returns it.
```

Comment:
I believe the better word here would be "Registers" instead of "Stores. Do I understand the intention right?